### PR TITLE
VertxThreadFactory service provider

### PR DESCRIPTION
**Pull request by vietj**

### CODE REVIEW

**Comment by vietj** [comment]
@geoand @FroMage this is the port of the thread customization to 4, can you have a look?

File: src/main/java/io/vertx/core/impl/VertxImpl.java [hunk 1]
```diff
@@ -1124,6 +1128,20 @@ public synchronized SharedWorkerPool createSharedWorkerPool(String name, int poo
...
 
+  private ThreadFactory createThreadFactory(long maxExecuteTime, TimeUnit maxExecuteTimeUnit, String prefix, boolean worker) {
+    AtomicInteger threadCount = new AtomicInteger(0);
+    return runnable -> {
+      VertxThread thread = threadFactory.newVertxThread(runnable, prefix + threadCount.getAndIncrement(), worker, maxExecuteTime, maxExecuteTimeUnit);
+      checker.registerThread(thread, thread);
+      // Vert.x threads are NOT daemons - we want them to prevent JVM exit so embededd user doesn't
+      // have to explicitly prevent JVM from exiting.
+      // I know the default is false anyway, but just to be explicit-  Vert.x threads are NOT daemons
+      // we want to prevent the JVM from exiting until Vert.x instances are closed
```
Comment (geoand) [inline src/main/java/io/vertx/core/impl/VertxImpl.java:1139]: Seems like the comment is duplicated :)

**Review by FroMage — APPROVED** [review]
LGTM, but this is a different API to the one we have in 3.x, right? So worth keeping in mind for https://github.com/quarkusio/quarkus/issues/14779

File: src/main/java/io/vertx/core/impl/VertxImpl.java (hunk 1 above)
Comment (geoand) [reply src/main/java/io/vertx/core/impl/VertxImpl.java:1139]: Also, should this really a lambda?
I don't know about Vert.x, but in Quarkus we try to stay away from lambdas as they have a slightly higher metaspace cost than regular classes.

**Review by geoand — APPROVED** [review]
I added a couple minor comments, but otherwise LGTM

**Comment by vietj** [comment]
yes it's different but it achieves the same right ?

**Comment by vietj** [comment]
if I would have done it myself for 3.x, that's the way I would have done it

**Comment by FroMage** [comment]
> yes it's different but it achieves the same right ?

Sure, but it's not binary compat, so one more thing for us to fix when we switch to 4.0.

> if I would have done it myself for 3.x, that's the way I would have done it

is it too late? was it released?

**Comment by vietj** [comment]
it has not been released, but it seems trivial

File: src/main/java/io/vertx/core/impl/VertxImpl.java (hunk 1 above)
Comment (vietj) [reply src/main/java/io/vertx/core/impl/VertxImpl.java:1139]: it's a lambda for readability, we have many lambda in other places and it seems it never raised an issue.

**Comment by vietj** [comment]
@FroMage I added a comment in the Quarkus 2 / Vert.x4 issue for max to record this change in the issue.

**Comment by FroMage** [comment]
Thanks.

**Pull request merged**